tests/nc-server/simple_server.c

```diff
@@ -0,0 +1,372 @@
+
+		/* Try to read some more data. */
+		if ((R->network_read_cookie = network_read(R->sock_read,
+		    R->buf, BUFLEN, 1, callback_read, R)) == NULL) {
+			warnp("network_read");
+			goto err0;
+		}
+	} else if (lenread == 0) {
+		if (drop(R)) {
+			warn0("drop");
+			goto err0;
+		}
+	} else {
+		warn0("Failed to read from network");
+		A->conndone = 1;
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Drop connection. */
+static int
+drop(struct conn_list_node * node_ptr)
+{
+
+	/* If we still have an active read cookie, cancel it. */
+	if (node_ptr->network_read_cookie != NULL)
+		network_read_cancel(node_ptr->network_read_cookie);
+
+	/* Close the incoming connection. */       
+	if (close(node_ptr->sock_read) == -1) {
+		warn0("close");
+		goto err0;
+	}
+
+	/* Clean up the node. */
+	conndied(node_ptr);
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * simple_server(port, nconn_max, shutdown_after, callback, caller_cookie):
+ * Run a server which accepts up to ${nconn_max} connections to port ${port}.
+ * After receiving a message, call ${callback} and pass it the
+ * ${caller_cookie}, along with the message.  Automatically shuts itself down
+ * after ${shutdown_after} connections have been dropped.
+ */
+int
+simple_server(uint16_t port, size_t nconn_max, size_t shutdown_after,
+    int (* callback_nc_message)(void *, uint8_t *, size_t),
+    void * caller_cookie)
+{
+	struct accept_state * A;
+	char * addr;
+	struct sock_addr ** sas;
+	int sock;
+
+	/* Create an address string suitable for sock_resolve. */
+	if (asprintf(&addr, "[0.0.0.0]:%" PRIu16, port) == -1) {
+		warn0("Out of memory");
+		goto err0;
+	}
+
+	/* Resolve the address. */
+	if ((sas = sock_resolve(addr)) == NULL) {
+		warn0("sock_resolve");
+		goto err1;
+	}
+
+	/* Create a socket, bind it, mark it as listening. */
+	if ((sock = sock_listener(sas[0])) == -1) {
+		warn0("sock_listener");
+		goto err2;
+	}
+
+	/* Bake a cookie for the server. */
+	if ((A = malloc(sizeof(struct accept_state))) == NULL) {
+		warnp("Out of memory");
+		goto err3;
+	}
+	A->s = sock;
+	A->conndone = 0;
+	A->shutdown_requested = 0;
+	A->shutdown_after = shutdown_after;
+	A->shutdown_current = 0;
+	A->nconn = 0;
+	A->nconn_max = nconn_max;
+	A->accept_cookie = NULL;
+	A->conn_cookies = NULL;
+	A->callback_nc_message = callback_nc_message;
+	A->caller_cookie = caller_cookie;
+
+	/* Accept a connection. */
+	if (doaccept(A)) {
+		warn0("doaccept");
+		goto err4;
+	}
+
+	/* Loop until we die. */
+	if (events_spin(&A->conndone)) {
+		warnp("Error running event loop");
+		goto err4;
+	}
+
+	/* Clean up. */
+	sock_addr_freelist(sas);
+	free(addr);
+	simple_server_shutdown(A);
+	events_shutdown();
+
+	/* Success! */
+	return (0);
+
+err4:
+	free(A);
+err3:
+	close(sock);
+err2:
+	sock_addr_freelist(sas);
+err1:
+	free(addr);
+	events_shutdown();
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
```

Comment:
Static functions go before non-static functions unless there's a very good reason to do otherwise.